File: gateway/internal/breaker/breaker.go
package breaker

import (
	"sync"
	"time"
)

// State represents the circuit breaker state.
type State int

const (
	Closed   State = iota // Normal: accepting frames
	Open                  // Tripped: rejecting frames
	HalfOpen              // Probing: allowing one frame to test recovery
)

func (s State) String() string {
	switch s {
	case Closed:
		return "closed"
	case Open:
		return "open"
	case HalfOpen:
		return "half-open"
	default:
		return "unknown"
	}
}

// Breaker is a circuit breaker that tracks iris-engine responsiveness.
//
// It opens (trips) when frames have been sent but no results received
// within the timeout window. It half-opens periodically to probe for recovery.
type Breaker struct {
	mu            sync.Mutex
	state         State
	timeout       time.Duration // How long without results before tripping
	probeInterval time.Duration // How often to probe when open
	lastSent      time.Time     // Last time a frame was sent
	lastResult    time.Time     // Last time a result was received
	lastProbe     time.Time     // Last time we allowed a probe frame
}

// New creates a circuit breaker with the given timeout and probe interval.
func New(timeout, probeInterval time.Duration) *Breaker {
	now := time.Now()
	return &Breaker{
		state:         Closed,
		timeout:       timeout,
		probeInterval: probeInterval,
		lastResult:    now, // Start as if we just got a result (don't trip on startup)
	}
}

// Allow checks whether a frame should be accepted.
// Returns true if the frame should be processed, false if rejected.
func (b *Breaker) Allow() bool {
	b.mu.Lock()
	defer b.mu.Unlock()

	now := time.Now()
	b.evaluate(now)

	switch b.state {
	case Closed:
		b.lastSent = now
		return true
	case HalfOpen:
		b.lastSent = now
		b.lastProbe = now
		b.state = Open // Back to open until we get a result
		return true
	case Open:
		return false
	}
	return false
}

// RecordResult signals that a result was received from iris-engine.
func (b *Breaker) RecordResult() {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.lastResult = time.Now()
	b.state = Closed
}

// State returns the current circuit breaker state.
func (b *Breaker) State() State {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.evaluate(time.Now())
	return b.state
}

// evaluate updates state based on timing (must be called with lock held).
func (b *Breaker) evaluate(now time.Time) {
	switch b.state {
	case Closed:
		// Trip if we've sent frames but got no results within timeout
		if !b.lastSent.IsZero() && b.lastSent.After(b.lastResult) &&
			now.Sub(b.lastSent) > b.timeout {
			b.state = Open
		}
	case Open:
		// Transition to half-open if enough time has passed since last probe
		if now.Sub(b.lastProbe) > b.probeInterval {
			b.state = HalfOpen
		}
	}
}

File: gateway/src/breaker.cpp
#include "breaker.h"

namespace eyed {

Breaker::Breaker(Duration timeout, Duration probe_interval)
    : state_(State::Closed),
      timeout_(timeout),
      probe_interval_(probe_interval),
      last_sent_(TimePoint{}),
      last_result_(TimePoint{}),
      last_probe_(TimePoint{}) {}

bool Breaker::allow() {
    std::lock_guard<std::mutex> lock(mu_);

    auto now = Clock::now();
    evaluate(now);

    switch (state_) {
        case State::Closed:
            last_sent_ = now;
            return true;

        case State::HalfOpen:
            last_sent_ = now;
            last_probe_ = now;
            state_ = State::Open;  // Back to open until we get a result
            return true;

        case State::Open:
            return false;
    }

    return false;
}

void Breaker::record_result() {
    std::lock_guard<std::mutex> lock(mu_);
    last_result_ = Clock::now();
    state_ = State::Closed;
}

State Breaker::state() {
    std::lock_guard<std::mutex> lock(mu_);
    evaluate(Clock::now());
    return state_;
}

std::string Breaker::state_string() const {
    std::lock_guard<std::mutex> lock(mu_);
    return eyed::state_string(state_);
}

void Breaker::evaluate(TimePoint now) {
    switch (state_) {
        case State::Closed:
            // Trip if we've sent frames but got no results within timeout
            if (last_sent_.time_since_epoch().count() != 0 &&
                last_sent_ > last_result_ &&
                (now - last_sent_) > timeout_) {
                state_ = State::Open;
                last_probe_ = now;  // Start probe interval from when we opened
            }
            break;

        case State::Open:
            // Transition to half-open if enough time has passed since last probe
            if ((now - last_probe_) > probe_interval_) {
                state_ = State::HalfOpen;
            }
            break;

        case State::HalfOpen:
            // No automatic transition from HalfOpen
            break;
    }
}

} // namespace eyed
